File: 2411-spiral-matrix-iv/2411-spiral-matrix-iv.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> spiralMatrix(int m, int n, ListNode* head) {
        int left = 0;
        int right = n;
        int top = 0;
        int bottom = m;

        vector<vector<int>> res(m, vector<int> (n, -1));
        
        while(left < right && top < bottom){
            for(int i=left; i<right; i++){
                res[top][i] = head->val;
                if (head->next == NULL){
                    return res;
                }
                head = head->next;
            }
            top++;

            for(int i=top; i<bottom; i++){
                res[i][right-1] = head->val;
                if (head->next == NULL){
                    return res;
                }
                head = head->next;
            }
            right--;

            for(int i=right-1; i>=left; i--){
                res[bottom-1][i] = head->val;
                if (head->next == NULL){
                    return res;
                }
                head = head->next;
            }
            bottom--;

            for(int i=bottom-1; i>=top; i--){
                res[i][left] = head->val;
                if (head->next == NULL){
                    return res;
                }
                head = head->next;
            }
            left++;
        }
        return res;
    }
};

File: 2411-spiral-matrix-iv/2411-spiral-matrix-iv.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def spiralMatrix(self, m: int, n: int, head: Optional[ListNode]) -> List[List[int]]:
        mat = [[-1 for i in range(n)] for _ in range(m)]
        
        left, right = 0, n
        top, bottom = 0, m
        while left < right and top < bottom:
            # put value in the top row
            for i in range(left, right):
                mat[top][i] = head.val
                if not head.next:
                    return mat
                head = head.next
            top += 1
            
            # put value in the right col
            for i in range(top, bottom):
                mat[i][right-1] = head.val
                if not head.next:
                    return mat
                head = head.next
            right -= 1
            
            for i in range(right-1, left-1, -1):
                mat[bottom-1][i] = head.val
                if not head.next:
                    return mat
                head = head.next
            bottom -= 1
            
            # put value in the left col
            for i in range(bottom-1, top-1, -1):
                mat[i][left] = head.val
                if not head.next:
                    return mat
                head = head.next
            left += 1
        return mat